tools/SwiftSyntax/SyntaxData.swift

```diff
@@ -198,3 +274,54 @@ final class SyntaxData: Equatable {
     return lhs === rhs
   }
 }
+
+/// An absolute position in a source file as text - the absolute offset from
+/// the start, line, and column.
+public class AbsolutePosition {
+  public private(set) var offset: Int = 0
+  public private(set) var line: Int = 1
+  public private(set) var column: Int = 1
+
+  /// Add some number of columns to the position.
+  internal func add(columns: Int) {
+    column += columns
+    offset += columns
+  }
+
+  /// Add some number of newlines to the position, resetting the column.
+  /// Size is byte size of newline char.
+  /// '\n' and '\r' are 1, '\r\n' is 2.
+  internal func add(lines: Int, size: Int) {
+    line += lines
+    column = 1
+    offset += lines * size
+  }
+
+  /// Use some text as a reference for adding to the absolute position,
+  /// taking note of newlines, etc.
+  internal func add(text: String) {
+    guard let chars = text.cString(using: .ascii) else {
```

Comment:
Is this going to skip all emoji or unicode characters?

Reply:
Good catch, we should use .utf8 here to handle emojis, etc.

Reply:
Well, that's not quite right either. `line` and `column` need to be grapheme-aware, i.e. 👩‍👩‍👧‍👧should not be 4 columns.

@milseman How do you think is the most efficient way to go about this? We want to track the number of characters and also the byte offset into the string here.

Reply:
hmm, the important factor here is to make the protocol consistent across multiple tools. SourceKit currently uses byte offset to interpret cursor-info (range selection) input. I think we should do the same here.

Reply:
Hmm, the best way requires more functionality in the standard library (that I really want to add anyways). We should support native UTF-8 and raw-initialization APIs. That way every String.Index has the byte offset into the original buffer, so index-to-source mappings are handled.

Reply:
Until then, you have to maintain the index to source location mapping yourself.